compiler.md

```diff
@@ -1,155 +1,483 @@
+  - [Videos](#videos)
+  - [More Resources](#more-resources)
 
 <!-- END doctoc generated TOC please keep comment here to allow auto update -->
 
-# v8 Compiler
+## Goals
 
-## Components
+[watch](https://youtu.be/HDuSEbLWyOY?t=7m22s)
 
-### Base Compiler 
+> Speed up real world performance for modern JavaScript, and enable developers to build a
+> faster future web.
 
-- is used for all code initially
-- generates code quickly without heavy optimizations
-- compilation with the base compiler is very fast generates little code
+- fast startup vs. peak performance
+- low memory vs. max optimization
+- Ignition Interpreter allows to run code with some amount of optimization very quickly and has
+  very low memory footprint
+- TurboFan makes functions that run a lot fast, sacrificing some memory in the process
+- designed to support entire JavaScript language and make it possible to quickly add new
+  features and to optimize them fast and incrementally
 
-### Runtime Profiler 
+## Simplified Pipeline
 
-- monitors the running system and identifies hot code
+[slide: pipeline 2010](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_163) |
+[slide: pipeline 2014](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_220) |
+[slide: pipeline 2016](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_249) |
+[slide: pipeline 2017](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_125)
 
-### Optimizing Compiler 
+Once crankshaft was taken out of the mix the below pipeline was possible
 
-- recompiles and optimizes hot code identified by the runtime profiler
-- uses static single assignment form to perform optimizations
-  - loop-invariant code motion
-  - linear-scan register allocation
-  - inlining.
-- optimization decisions are based on type information collected while running the code produced by the base compiler
+<img alt="simplified pipeline" src="http://benediktmeurer.de/images/2016/v8-new-pipeline-20161125.png" width="70%">
 
-### Deoptimization Support 
+### Basic Steps
 
-- allows the optimizing compiler to be optimistic in the assumptions it makes when generating code
-- deoptimization support allows to bail out to the code generated by the base compiler if the assumptions in the
-  optimized code turn out to be too optimistic
+1. Parse JavaScript into an [AST (abstract syntax tree)](https://en.wikipedia.org/wiki/Abstract_syntax_tree)
+2. Generate bytecode from that AST
+3. Turn bytecode into sequence of bytecodes by the BytecodeGenerator, which is part of the [Ignition Interpreter](https://v8project.blogspot.com/2016/08/firing-up-ignition-interpreter.html)
+  - sequences are divided on a per function basis
+4. Execute bytecode sequences via Ignition and collect feedback via inline caches
```

Comment:
(e.g. "x" is a variable that has always been holding a string)

Reply:
This is a summary of the basic steps.
The details are explained in [this section](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#collecting-feedback-via-ics) including a feedback lattice.